v2/pkg/astnormalization/uploads/upload_finder.go

```diff
@@ -0,0 +1,270 @@
+package uploads
+
+import (
+	"bytes"
+
+	"github.com/wundergraph/astjson"
+
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/ast"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/internal/unsafebytes"
+)
+
+var (
+	uploadScalarName = []byte("Upload")
+	variablesLiteral = []byte("variables")
+)
+
+type UploadFinder struct {
+	operation                 *ast.Document
+	definition                *ast.Document
+	variables                 *astjson.Value
+	currentVariableName       []byte
+	currentVariableValue      *astjson.Value
+	currentArgPath            *Path
+	currentVariablePath       *Path
+	variableUsedDirectlyOnArg bool
+
+	uploadPathMapping []UploadPathMapping
+}
+
+type UploadPathMapping struct {
+	VariableName       string
+	OriginalUploadPath string
+	NewUploadPath      string
+}
+
+func NewUploadFinder() *UploadFinder {
+	return &UploadFinder{
+		currentArgPath:      &Path{},
+		currentVariablePath: &Path{},
+	}
+}
+
+func (v *UploadFinder) Reset() {
+	v.operation = nil
+	v.definition = nil
+	v.variables = nil
+	v.currentVariableName = nil
+	v.currentVariableValue = nil
+	v.currentArgPath.reset()
+	v.currentVariablePath.reset()
+	v.variableUsedDirectlyOnArg = false
+	v.uploadPathMapping = v.uploadPathMapping[:0]
+}
+
+func (v *UploadFinder) FindUploads(operation, definition *ast.Document, variables []byte, argRef int, argInputValueDefinitionRef int) (uploadPathMapping []UploadPathMapping, err error) {
+	v.definition = definition
+	v.operation = operation
+
+	if variables == nil || bytes.Equal(variables, []byte("null")) || bytes.Equal(variables, []byte("")) {
+		variables = []byte("{}")
+	}
+
+	v.variables, err = astjson.ParseBytesWithoutCache(variables)
+	if err != nil {
+		return nil, err
+	}
+
+	v.currentArgPath.reset()
+	v.currentVariablePath.reset()
+
+	argumentValue := v.operation.Arguments[argRef].Value
+	argumentTypeRef := v.definition.InputValueDefinitionType(argInputValueDefinitionRef)
+
+	v.traverseValue(argumentValue, argumentTypeRef)
+
+	if len(v.uploadPathMapping) > 0 {
+		return v.uploadPathMapping, nil
+	}
+
+	return nil, nil
+}
+
+func (v *UploadFinder) traverseValue(value ast.Value, valueTypeRef int) {
+	switch value.Kind {
+	case ast.ValueKindVariable:
```

Comment:
Can you add more comments for each case to explain why and how you handle them? e.g Variables